keps/sig-node/4438-sidecar-restart-termination/README.md

```diff
@@ -350,9 +356,53 @@ For Beta and GA, add links to added tests together with links to k8s-triage for
--->
 
 - <test>: <link to test coverage>
+-->
+
+###### Existing tests
+
+- should respect termination grace period seconds
+- should respect termination grace period seconds with long-running preStop hook https://github.com/kubernetes/kubernetes/blob/fbb2e6293fb0c8c107ae48b8b8ae488325c59598/test/e2e_node/container_lifecycle_test.go#L536
+- should call the container's preStop hook and terminate it if its startup probe fails https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/container_lifecycle_test.go#L616
+- should call the container's preStop hook and terminate it if its liveness probe fails https://github.com/kubernetes/kubernetes/blob/fbb2e6293fb0c8c107ae48b8b8ae488325c59598/test/e2e_node/container_lifecycle_test.go#L683
+
+###### New tests
+
+Probes:
+- Readiness probes are still running while in preStop
+- Readiness status is beings updated for the container and the Pod while in preStop
+- Liveness probes are NOT running for regular containers while the Pod is terminating 
+- SIDECAR: Liveness probes DO run for sidecar containers while the Pod is terminating 
+- SIDECAR: sidecar container will be restarted when liveness probe failed during Pod termination
+
+Not fully started containers:
+- preStop will not be executed for the container that hasn’t started yet
+- preStop will be called on the container even if postStart is still running
+- postStart hook CONTINUE EXECUTE even if container started termination
+- postStart hook will stop once pod passed it’s graceful termination period
+
+Pod with some containers terminated:
+- BUGFIX, SIDECAR: Container can be restarted when there are terminating containers in the Pod A container cannot restart when there is any terminating container in the same pod · Issue #121398
+
+Re-terminating the Pod:
+- When the Pod is terminating, another call to terminate the pod with the smaller grace period will override the grace period to terminate Pod faster
+- When the Pod is terminating, another call to terminate the pod with the greater grace period will override the grace period to allow longer termination
+- BUGFIX: Service account token gets invalidated while terminating pod is re-deleted · Issue #122568
+
+Pre-stop vs. SIGTERM traps:
+- Same as existing and above tests, need to validate that the container that traps the SIGTERM behaves the same way as with preStop:
+- Respect the grace period
+- Liveness probes are not running
+- Readiness probes are running
+
+Test what is available for during preStop:
+- BUGFIX: While the Pod is terminating, service account tokens are rotated Kubelet stops rotating service account tokens when pod is terminating, breaking preStop hooks · Issue #116481
+- BUGFIX: Service account token is valid if the terminating Pod was deleted again Service account token gets invalidated while terminating pod is re-deleted · Issue #122568
+
+Eviction and OOM kills:
+- preStop is called when Pod is evicted
+- preStop is NOT called when Container is OOMkilled
```

Comment:
Commenting here since I can't comment on unchanged lines:

1. Please fill-in Upgrade/downgrade strategy and Vesion skew

Should be fairly simple for this feature since this only touches kubelet, but we need it for completeness.


2. Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

I somewhat disagree with the answer - given that I don't need to change anything in the pod, it actually will affect the running workloads (when they will be terminating), right?


3.  Are there any tests for feature enablement/disablement?

If you want to target beta, can you be more specific what those tests will be?


4. If you want to target beta directly, then the rest of the PRR should be filled in.

Reply:
thanks @wojtek-t , updated with hopefully all remarks